### ISSUE TITLE
Buggy register liveness detection, rdi gets clobbered

### ISSUE THREAD

**Issue by kaott** [dyninst/dyninst#724]
**Intention**
The main intention is to insert instrumentation code at both the entry point of a function and the exit point of a function.  It only seems to be a problem in the instrumentation code that is inserted at the end of the function. Seems to be similar to issue #434.

**Describe the bug**
The instrumentation code involves some syscalls which make use of the rdi register, this register is unfortunately not saved properly and gets filled with an a syscall parameter. Later in the code the rdi contents of the rdi register are used as a memory address, but the contents are garbage so the program segfaults. If I disable the liveness analysis the newly mutated program works properly. Listed below is the relevant assembly.
```
...
  50ad3b:       50                      push   %rax
  50ad3c:       51                      push   %rcx
  50ad3d:       52                      push   %rdx
  50ad3e:       48 8d 64 24 f8          lea    -0x8(%rsp),%rsp
  50ad43:       48 b8 00 00 00 00 00    movabs $0x0,%rax
  50ad4a:       00 00 00
  50ad4d:       ff 15 45 b8 ff ff       callq  *-0x47bb(%rip)        # 506598 <fg_perf_stop@Base>
  50ad53:       48 bf 24 00 00 00 00    movabs $0x24,%rdi
  50ad5a:       00 00 00
  50ad5d:       48 b8 00 00 00 00 00    movabs $0x0,%rax
  50ad64:       00 00 00
  50ad67:       ff 15 33 b8 ff ff       callq  *-0x47cd(%rip)        # 5065a0 <fg_perf_read@Base>
  50ad6d:       48 b8 00 00 00 00 00    movabs $0x0,%rax
  50ad74:       00 00 00
  50ad77:       ff 15 2b b8 ff ff       callq  *-0x47d5(%rip)        # 5065a8 <fg_perf_start@Base>
  50ad7d:       48 8d 64 24 08          lea    0x8(%rsp),%rsp
  50ad82:       5a                      pop    %rdx
  50ad83:       59                      pop    %rcx
  50ad84:       58                      pop    %rax
  50ad85:       48 8b 24 24             mov    (%rsp),%rsp
...
```

The calls in fg_perf_* use ioctls syscalls which ends up clobbering the rdi register. The following code gets called later, which rdi is properly saved but it is too late as it contains garbage contents, which results in a segfault because the garbage contents are then dereferenced.
```
...
  50ac6e:       5f                      pop    %rdi
  50ac6f:       5e                      pop    %rsi
  50ac70:       59                      pop    %rcx
  50ac71:       41 58                   pop    %r8
  50ac73:       48 8b 24 24             mov    (%rsp),%rsp
  50ac77:       48 8b 07                mov    (%rdi),%rax
...
```

**To Reproduce**
It is directly reproducible on most of the programs in my test bench. I can include any other requested information.

**Expected behavior**
The contents of the rdi register are saved and restored properly and the mutated program doesn't segfault.

**System (please complete the following information):**
x86_64
DynInst version 10.1
glibc 2.30, gcc 9.2

**Comment by sashanicolas** [comment]
@hainest do we have a system with libc 2.30?

**Comment by hainest** [comment]
@sashanicolas No. Ubuntu 19.10 has it, but you'll need to install it in a VM.

**Comment by hainest** [comment]
@kaott Could you provide a reproducer either in source or binary form?

**Comment by kaott** [comment]
Here is a smallish test bed that reproduces the problem. Included is a driver program for DynInst, a makefile, the mutator library, and a sample binary I am looking to modify.
[files_rdi.tar.gz](https://github.com/dyninst/dyninst/files/4703351/files_rdi.tar.gz)

The makefile will need to be updated to point to the correct paths for the DynInst install location. 

The following environment variables are also set:
`export LD_LIBRARY_PATH=/path/to/libfgperf/in/tarball`
`export DYNINSTAPI_RT_LIB=/path/to/libdyninstAPI_RT.so`

Let me know if you need anything else.
